### ISSUE TITLE
confusing docstring for Problem.solve

### ISSUE THREAD

**Issue by keithbriggs** [cvxgrp/cvxpy#890]
it says:
    solver_specific_opts (dict, optional) – A dict of options that will be passed to the specific solver. 

Does this mean the whole dict is passed, or is it unpacked and the separate arguments passed?  E.g., if I write

      d={'abstol': 1e-4}
      sol=prob.solve(solver=cp.ECOS,solver_specific_opts=d,verbose=True)

does it pass an argument "abstol=1e-4" to ECOS?  (It seems that it does not.)

In general, where are the solver-specific options documented?

**Comment by SteveDiamond** [comment]
The solver specific options are documented here: https://www.cvxpy.org/tutorial/advanced/index.html

**Comment by rileyjmurray** [comment]
@SteveDiamond I believe the issue is that the docstring of ``prob.solve`` may not be accurate. It says there is a specific keyword argument called ``solver_specific_opts``, which is dictionary-valued. From what I can tell, cvxpy doesn't look for a keyword argument by that name.

There are a few solvers which accept dictionary-valued keyword arguments: NAG, CPLEX, and MOSEK. However those solvers don't check for a key named ``solver_specific_opts``, they check for keys of the form ``[SOLVER]_params`` (where ``[SOLVER]`` should be replaced by ``mosek`` or ``cplex`` or ``nag``).

I think it would be good to change the docstring of ``prob.solve`` to remove reference to ``solver_specific_opts``. The docstring should just have a "notes" section, where it explains how solver specific options are passed in. For example ...
```
Notes
------
CVXPY interfaces with a wide range of solvers; the algorithms used by these solvers
have parameters relating to stopping criteria, and strategies to improve solution quality.
    
There is no one choice of parameters which is perfect for every problem. If you are not
getting satisfactory results from a solver, you can try changing its parameters. The
exact way this is done depends on the specific solver. Here are some examples:

    prob.solve(solver='ECOS', abstol=1e-6)
    prob.solve(solver='OSQP', max_iter=10000).
    mydict = {"MSK_DPAR_INTPNT_CO_TOL_NEAR_REL":  10}
    prob.solve(solver='MOSEK', mosek_params=mydict).

You should refer to CVXPY's web documentation for details on how to pass solver
parameters.

```

**Comment by SteveDiamond** [comment]
I see, let's change the docstring. I can do it once your PR is merged in. I don't want to create more merge problems.

**Comment by SteveDiamond** [comment]
Fixed.

**Comment by iarbel84** [comment]
@SteveDiamond how should I pass the parameters? Specifically, the tolerance. I tried for Gurobi, ECOS, Mosek. It looks like non of them actually uses the parameter I pass. For Gurobi I put `OptimalityTol=1e-4`, for ECOS `abstol=1e-4` and Mosek got `mosek_params={'MSK_DPAR_INTPNT_CO_TOL_REL_GAP': 1e-4}`. All solvers ignore this param and give a 1e-8 default solution. I'm clueless as to what to do

**Comment by rileyjmurray** [comment]
@iarbel84 the way that solvers check and handle termination criteria can be a bit inconsistent with what users expect. As an example, I created a problem representing a "power cone program" -- the official cvxpy repo doesn't yet support these, but they are planned in the next PR. So let's say this problem object is called ``prob``. Here are two examples of calling SCS with different solver tolerances:
```
prob.solve(solver='SCS', verbose=True, eps=1e-4)

----------------------------------------------------------------------------
	SCS v2.1.2 - Splitting Conic Solver
	(c) Brendan O'Donoghue, Stanford University, 2012
----------------------------------------------------------------------------
Lin-sys: sparse-direct, nnz in A = 8
eps = 1.00e-04, alpha = 1.50, max_iters = 5000, normalize = 1, scale = 1.00
acceleration_lookback = 10, rho_x = 1.00e-03
Variables n = 5, constraints m = 7
Cones:	primal zero / dual free vars: 1
	primal + dual power vars: 6
Setup time: 5.34e-04s
----------------------------------------------------------------------------
 Iter | pri res | dua res | rel gap | pri obj | dua obj | kap/tau | time (s)
----------------------------------------------------------------------------
     0| 1.40e+19  1.98e+19  1.00e+00 -4.45e+19  5.59e+19  2.26e+19  2.58e-04 
    20| 2.57e-08  3.94e-08  2.96e-09 -1.81e+00 -1.81e+00  1.86e-16  7.96e-04 
----------------------------------------------------------------------------
Status: Solved
Timing: Solve time: 9.17e-04s
	Lin-sys: nnz in L factor: 20, avg solve time: 1.94e-07s
	Cones: avg projection time: 1.34e-06s
	Acceleration: avg step time: 2.02e-05s
----------------------------------------------------------------------------
Error metrics:
dist(s, K) = 2.0225e-17, dist(y, K*) = 0.0000e+00, s'y/|s||y| = -6.3116e-17
primal res: |Ax + s - b|_2 / (1 + |b|_2) = 2.5671e-08
dual res:   |A'y + c|_2 / (1 + |c|_2) = 3.9380e-08
rel gap:    |c'x + b'y| / (1 + |c'x| + |b'y|) = 2.9618e-09
----------------------------------------------------------------------------
c'x = -1.8073, -b'y = -1.8073
============================================================================
```
note that the solver output clearly states ``eps = 1.00e-04``. So, cvxpy correctly passed this parameter. We can try again, with explicit tolerance of 1e-8. We see that the solver log prints ``eps = 1.00e-08``. The final solution ends up being accurate to 1e-10.
```
prob.solve(solver='SCS', verbose=True, eps=1e-8)

----------------------------------------------------------------------------
	SCS v2.1.2 - Splitting Conic Solver
	(c) Brendan O'Donoghue, Stanford University, 2012
----------------------------------------------------------------------------
Lin-sys: sparse-direct, nnz in A = 8
eps = 1.00e-08, alpha = 1.50, max_iters = 5000, normalize = 1, scale = 1.00
acceleration_lookback = 10, rho_x = 1.00e-03
Variables n = 5, constraints m = 7
Cones:	primal zero / dual free vars: 1
	primal + dual power vars: 6
Setup time: 9.83e-04s
----------------------------------------------------------------------------
 Iter | pri res | dua res | rel gap | pri obj | dua obj | kap/tau | time (s)
----------------------------------------------------------------------------
     0| 1.40e+19  1.98e+19  1.00e+00 -4.45e+19  5.59e+19  2.26e+19  3.55e-04 
    25| 6.40e-12  1.04e-11  4.77e-12 -1.81e+00 -1.81e+00  3.04e-16  1.94e-02 
----------------------------------------------------------------------------
Status: Solved
Timing: Solve time: 1.96e-02s
	Lin-sys: nnz in L factor: 20, avg solve time: 1.14e-06s
	Cones: avg projection time: 2.45e-06s
	Acceleration: avg step time: 7.26e-04s
----------------------------------------------------------------------------
Error metrics:
dist(s, K) = 3.2031e-17, dist(y, K*) = 0.0000e+00, s'y/|s||y| = -6.3116e-17
primal res: |Ax + s - b|_2 / (1 + |b|_2) = 6.3978e-12
dual res:   |A'y + c|_2 / (1 + |c|_2) = 1.0420e-11
rel gap:    |c'x + b'y| / (1 + |c'x| + |b'y|) = 4.7686e-12
----------------------------------------------------------------------------
c'x = -1.8073, -b'y = -1.8073
```

For ECOS, there is [an outstanding issue](https://github.com/embotech/ecos/issues) on their GitHub (issue 156, which I opened in late 2017) about handling stopping criteria. That doesn't rule out a bug in our ECOS interface, but it serves to illustrate that solver tolerance checks can be different from what we (as users) might expect.

For MOSEK, I once found that solver behavior wasn't changing when I was selecting extreme values that should affect algorithm performance. So I emailed MOSEK support about this, and they told me there was actually a bug in their code where they didn't process the user-provided value. In this MOSEK case, the parameter I changed was unusual, so it makes sense that the bug in question flew under the radar. In your case, ``CO_TOL_REL_GAP`` is an important parameter, so it's unlikely that MOSEK isn't processing this parameter. Perhaps there *is* a bug in CVXPY. To see if we're processing tolerance parameters correctly, try also passing a maximum number of interior point iterations, e.g.,
```
mosek_params = {
                mosek.dparam.basis_tol_x: 1e-9,
                "MSK_IPAR_INTPNT_MAX_ITERATIONS": 6
            }
```
Let me know how that goes.

**Comment by iarbel84** [comment]
@rileyjmurray thanks for the detailed reply! I wasn't sure if the problem was with the solvers or with CVXPY... since it didn't work for 3 solvers I thought it's CVXPY or that I'm doing something wrong. BTW, I also passed number of iterations as a parameter and that one worked. In any case, based on what you wrote it looks like the problem is from the solvers, and since getting a fast (but not very precise) solution is crucial to the sequence of problems I'm solving, I used decided to explore other options. It turned out that an accelerated first-order method (FISTA) was much faster than these IPM solvers. It actually performed x10 times faster than MOSEK which was the best solver, and it doesn't have the scale limit that IPM algorithms have
